src/csharp/Grpc.Core/ChannelCredentials.cs

```diff
@@ -104,40 +104,71 @@ internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
+    /// <param name="targetPem">string containint PEM encoded certificate of the peer.</param>
+    /// <returns>0 is verification succeeded, non 0 otherwise.</returns>
+    public delegate bool VerifyPeerCallback(string targetHost, string targetPem);
+
+    internal delegate int VerifyPeerCallbackInternal(
+        string targetHost,
+        string targetPem,
+        IntPtr userData);
+
     /// <summary>
     /// Client-side SSL credentials.
     /// </summary>
     public sealed class SslCredentials : ChannelCredentials
     {
         readonly string rootCertificates;
         readonly KeyCertificatePair keyCertificatePair;
+        readonly VerifyPeerCallbackInternal verifyPeerCallback;
 
         /// <summary>
         /// Creates client-side SSL credentials loaded from
         /// disk file pointed to by the GRPC_DEFAULT_SSL_ROOTS_FILE_PATH environment variable.
         /// If that fails, gets the roots certificates from a well known place on disk.
         /// </summary>
-        public SslCredentials() : this(null, null)
+        public SslCredentials() : this(null, null, null)
         {
         }
 
         /// <summary>
         /// Creates client-side SSL credentials from
         /// a string containing PEM encoded root certificates.
         /// </summary>
-        public SslCredentials(string rootCertificates) : this(rootCertificates, null)
+        public SslCredentials(string rootCertificates) :
+            this(rootCertificates, null, null)
         {
         }
-            
+
+        /// <summary>
+        /// Creates client-side SSL credentials.
+        /// </summary>
+        /// <param name="rootCertificates">string containing PEM encoded server root certificates.</param>
+        /// <param name="keyCertificatePair">a key certificate pair.</param>
+        public SslCredentials(string rootCertificates, KeyCertificatePair keyCertificatePair) :
+            this(rootCertificates, keyCertificatePair, null)
+        {
+        }
+
         /// <summary>
         /// Creates client-side SSL credentials.
         /// </summary>
         /// <param name="rootCertificates">string containing PEM encoded server root certificates.</param>
         /// <param name="keyCertificatePair">a key certificate pair.</param>
-        public SslCredentials(string rootCertificates, KeyCertificatePair keyCertificatePair)
+        /// <param name="verifyPeerCallback">a callback to verify peer's target name and certificate.</param>
+        public SslCredentials(string rootCertificates, KeyCertificatePair keyCertificatePair, VerifyPeerCallback verifyPeerCallback)
         {
             this.rootCertificates = rootCertificates;
             this.keyCertificatePair = keyCertificatePair;
+            this.verifyPeerCallback = verifyPeerCallback == null ?
+                null : new VerifyPeerCallbackInternal((host, pem, userData) => verifyPeerCallback(host, pem) ? 0 : 1);
```

Comment:
this is actually not safe - the problem is that the delegate can be garbage collected at some point but the native layer won't know about it and will try to invoke it, which will result in segfault.
Native to managed callbacks are tricky.
Look at https://github.com/grpc/grpc/blob/3278bdceda8030d5aa130f12765e5f07263c860d/src/csharp/Grpc.Core/Internal/NativeMetadataCredentialsPlugin.cs#L47 to see how the destruction logic is handled (it's not that simple).

Reply:
Thank you for pointing me to this. I had sort of guessed that there could be some issue with the callback. I first added a `GC.Collect()` in one of the test to highlight the issue (https://github.com/grpc/grpc/pull/17051/files#diff-576c954e2167da66d782732e2a4196d7R226) and then implemented a fix for it. I would have like the fix to be less impactful but could not really find a better way than what I have done now. The main issue is that the channel does not keep track of the channel credentials that are passed in its constructor. A naive way would have been to simply add a reference to it, but I thought that only keeping track of the `GCHandle` when necessary would be slightly cleaner. Let me know what you think.